### DISCUSSION TITLE
type of a declaratively mapped class

### DISCUSSION THREAD

**Discussion by cjw296** [sqlalchemy/sqlalchemy#8623] [Usage Questions]
If I have a class like this:

```
class Simple(Base):
    __tablename__ = 'simple'
    key = Column(String, primary_key=True)
    value = Column(Integer, unique=True)
```
Is there a generic way of expressing `Type[Base]` in a [library that works with any such objects](https://github.com/Mortar/mortar_import/blob/ca07fd9f67cab66a6a2020932fd05b3ffcd20fbe/mortar_import/sqlalchemy.py#L18-L22)?
I guess I'm looking for a type annotation that says "must be a declaratively mapped class"...

**Comment by CaselIT** [comment]
Hi,

I don't think there is in general a superclass that you can assume is used as base in all models, since sqlalchemy supports mapping classes that are not subclass of a particular base.

A general `type` would be the best type I think.

Maybe slqlachemy could provide a Protocol for the Models, but I don't know what attributes there could be there. Maybe `__table__`?
Thought on this @zzzeek ?

**Reply by zzzeek** [reply]
we have this in 2.0 using [DeclarativeBase](https://github.com/sqlalchemy/sqlalchemy/blob/fd485536664967d5ffad34b3184e7b63d0816d39/lib/sqlalchemy/orm/decl_api.py#L610).  however, that only counts for models that were mapped using that specific technique, it won't cover models that were mapped with `@registry.mapped` or `@registry.map_imperatively`.

**Reply by CaselIT** [reply]
yes, I almost suggested it, but as mentioned it doesn't cover all cases.
That's the reason why I suggested a protocol.

The protocol could probably just copy the same data that's in the type checking block of `DeclarativeBase`
Alternately if making `DeclarativeBase` a Protocol itself does not interfere with sqlalchemy or type checking in general maybe that could also be an option?

**Reply by zzzeek** [reply]
There's no "protocol" for a classically mapped class, so I dont see how that changes the situation much; mapping is inherently a "transparent" implementation detail that does not have any implications for typing by itself (Edit: OK i forgot we [add these now](https://docs.sqlalchemy.org/en/14/orm/mapping_styles.html#runtime-introspection-of-mapped-classes-instances-and-mappers)).      DeclarativeBase is the standard in 2.0 for declarative classes so that would be the "protocol" if one wanted to standardize on `__table__` etc. 

i sort of thought we had a protocol lying around for the `__mapper__`, `__table__` thing described at https://docs.sqlalchemy.org/en/14/orm/mapping_styles.html#runtime-introspection-of-mapped-classes-instances-and-mappers .

**Reply by CaselIT** [reply]
So there could be a protocol with the things that are added every time. so `__table__`  and `__mapper__`?

**Reply by CaselIT** [reply]
Any suggestion on the name? `MappedClass`?

**Reply by zzzeek** [reply]
sure that's pretty much it

**Reply by CaselIT** [reply]
Ok. I'll open an issue with this (mainly docs I guess)

**Reply by CaselIT** [reply]
the issue is this one https://github.com/sqlalchemy/sqlalchemy/issues/8624

**Comment by cjw296** [comment]
@zzzeek - "we have this in 2.0" confuses me: How do I use 2.0? I'm doing the `future=True` stuff where I can, but it's unclear if that's what you mean and I don't see any 2.0.devX releases on https://pypi.org/project/SQLAlchemy/#history

**Reply by CaselIT** [reply]
that's because it's version 2 not yet released. A first beta should be out this month!

**Reply by cjw296** [reply]
🤞

**Reply by zzzeek** [reply]
https://www.sqlalchemy.org/download.html#relstatus

**Reply by cjw296** [reply]
Right, so the "we have this in 2.0" is just teasing at this point? ;-)
